src/infer.py

```diff
@@ -0,0 +1,111 @@
+from imageio import imwrite
+from fridge import load_model
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string("checkpoint", None, "Path to checkpoint file.")
+flags.mark_flag_as_required("checkpoint")
+flags.DEFINE_string("numfonts", None, "Number of fonts to generate.")
+flags.mark_flag_as_required("numfonts")
+flags.DEFINE_string("path", None, "Path to save pngs to.")
+flags.mark_flag_as_required("path")
+
+CHARACTERS = [
+    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O',
+    'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd',
+    'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's',
+    't', 'u', 'v', 'w', 'x', 'y', 'z', 'zero', 'one', 'two', 'three', 'four',
+    'five', 'six', 'seven', 'eight', 'nine', 'exclam', 'numbersign', 'dollar',
+    'percent', 'ampersand', 'asterisk', 'question', 'at'
+]
+
+DEVICE = torch.device("cuda:1" if torch.cuda.is_available() else "cpu")
+
+
+def rasterize(x, resolution=64, sigma=0.01):
+    '''
+    Simple rasterization: drop a single Gaussian at every control point.
+
+    Parameters
+    ----------
+    x : [batch_size, num_control_points, 2]
+        Control points of glyphs.
+
+    Notes
+    -----
+    The num_contours and num_beziers dimensions have been collapsed into one
+    num_control_points dimension.
+
+    Also, we can pad with sufficiently large coordinates (e.g. 999) to
+    indicate that there are no more control points: this places a Gaussian
+    off-raster, which minimally affects the raster.
+    '''
+
+    # Padding constants chosen by looking at empirical distribution of
+    # coordinates of Bezier control point from real fonts
+    left_pad = 0.25 * resolution
+    right_pad = 1.25 * resolution
+    up_pad = 0.8 * resolution
+    down_pad = 0.4 * resolution
+    mesh_lr = np.linspace(
+        -left_pad, resolution + right_pad, num=resolution, endpoint=False)
+    mesh_ud = np.linspace(
+        -down_pad, resolution + up_pad, num=resolution, endpoint=False)
+    XX, YY = np.meshgrid(mesh_lr, mesh_ud)
+    YY = np.flip(YY)
+    XX_expanded = XX[:, :, np.newaxis]
+    YY_expanded = YY[:, :, np.newaxis]
+    x_meshgrid = torch.Tensor(XX_expanded / resolution).to(device)
+    y_meshgrid = torch.Tensor(YY_expanded / resolution).to(device)
+
+    batch_size = x.size()[0]
+    num_samples = x.size()[1]
+    x_samples = x[:, :, 0].unsqueeze(1).unsqueeze(1)
+    y_samples = x[:, :, 1].unsqueeze(1).unsqueeze(1)
+
+    x_meshgrid_expanded = x_meshgrid.expand(batch_size, resolution, resolution,
+                                            num_samples)
+    y_meshgrid_expanded = y_meshgrid.expand(batch_size, resolution, resolution,
+                                            num_samples)
+
+    raster = torch.exp((-(x_samples - x_meshgrid_expanded)**2 -
+                        (y_samples - y_meshgrid_expanded)**2) / (2 * sigma**2))
+    raster = raster.sum(dim=3)
+    return raster
+
+
+def save_images(images, size, path):
```

Comment:
best put in this in a util.py type of thing b/c the periodic val in the train script will use it too